File: in_place.py
def quicksort(a, lo=0, hi=None):
    if hi is None:
        hi = len(a) - 1
    if lo >= hi:
        return

    p = partition(a, lo, hi)
    quicksort(a, lo, p - 1)
    quicksort(a, p + 1, hi)


def partition(a, lo, hi):
    pivot = a[hi]
    i = lo
    for j in range(lo, hi):
        if a[j] < pivot:
            a[i], a[j] = a[j], a[i]
            i += 1
    a[i], a[hi] = a[hi], a[i]
    return i


if __name__ == "__main__":
    import sys
    verbose = '-v' in sys.argv
    pre_print = '-p' in sys.argv
    xs = [(i * 73 + 19) % 10000 for i in range(10000)]
    if pre_print:
        print('Before sorting:')
        print(xs)
        sys.exit(0)
    quicksort(xs)
    if verbose:
        print(xs)
    else:
        print(xs[:10])

File: in_place.js
function partition(arr, lo, hi) {
    const pivot = arr[hi];
    let i = lo;
    
    for (let j = lo; j < hi; j++) {
        if (arr[j] < pivot) {
            [arr[i], arr[j]] = [arr[j], arr[i]];
            i++;
        }
    }
    
    [arr[i], arr[hi]] = [arr[hi], arr[i]];
    return i;
}

function qsort(arr, lo, hi) {
    if (lo >= hi) {
        return;
    }
    
    const p = partition(arr, lo, hi);
    qsort(arr, lo, p - 1);
    qsort(arr, p + 1, hi);
}

function quicksort(arr) {
    if (arr.length > 0) {
        qsort(arr, 0, arr.length - 1);
    }
}

function main() {
    const args = process.argv.slice(2);
    const verbose = args.includes('-v');
    const prePrint = args.includes('-p');
    
    // Generate test data
    const xs = Array.from({length: 10000}, (_, i) => (i * 73 + 19) % 10000);
    
    if (prePrint) {
        console.log('Before sorting:');
        console.log(JSON.stringify(xs));
        return;
    }
    
    quicksort(xs);
    
    if (verbose) {
        console.log(JSON.stringify(xs));
    } else {
        console.log(JSON.stringify(xs.slice(0, 10)));
    }
}

main();
